### PR TITLE
Trend Micro Deep Security Integration

### PR DESCRIPTION
**Pull request by RazOmry**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Trend Micro Deep Security is a system that provides a runtime security solution for endpoints and workloads in the cloud.

The Deep Security API enables you to automate operational tasks, thereby increasing the productivity of your value stream and improving the security services that you support:
- Configure policies and protect computers.
- Discover vulnerabilities and patch them.
- Perform routine maintenance tasks.
- And more!

## Minimum version of Cortex XSOAR
- [x] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 1]
```diff
@@ -0,0 +1,1824 @@
...
+        Args:
+            expand (List[str]): The desired information about the computers.
+            overrides (bool): Whether to get the overridden properties or not.
+
+        Returns:
+            List[Dict[str, Any]]: The list of all computers.
+        """
+
+        params = {"expand": expand, "overrides": overrides}
+        return self._http_request(method="GET", url_suffix="/computers", params=params)["computers"]
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:53]:
```suggestion
        return self._http_request(method="GET", url_suffix="/computers", params=params).get("computers")
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 2]
```diff
@@ -0,0 +1,1824 @@
...
+        Args:
+            computer_id (int): The ID of the computer get its rules.
+            overrides (bool): Whether to get the rule IDs assigned directly to the current computer or not.
+
+        Returns:
+            List[int]: The rule IDs.
+        """
+
+        return self._http_request(method="GET", url_suffix=f"/computers/{computer_id}/firewall/assignments",
+                                  params={"overrides": overrides})["assignedRuleIDs"]
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:178]:
```suggestion
                                  params={"overrides": overrides}).get("assignedRuleIDs")
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 3]
```diff
@@ -0,0 +1,1824 @@
...
+            computer_id (int): The computer ID to assign the rules.
+            rule_ids (List[int]): The IDs of the rules to assign to the computer.
+            overrides (bool): Whether to get the rule IDs assigned directly to the current computer or not.
+
+        Returns:
+            List[int]: All rule IDs that are assigned to the computer.
+        """
+
+        return self._http_request(method="PUT", url_suffix=f"/computers/{computer_id}/firewall/assignments",
+                                  params={"overrides": overrides}, json_data={"rule_ids": rule_ids})["assignedRuleIDs"]
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:210]:
```suggestion
                                  params={"overrides": overrides}, json_data={"rule_ids": rule_ids}).get("assignedRuleIDs")
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 4]
```diff
@@ -0,0 +1,1824 @@
...
+
+    def list_computer_groups(self) -> List[Dict[str, Any]]:
+        """
+        List all computer groups inside Trend Micro.
+
+        Returns:
+            List[Dict[str, Any]]: All existing computer groups.
+        """
+
+        return self._http_request(method="GET", url_suffix="/computergroups")["computerGroups"]
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:232]:
```suggestion
        return self._http_request(method="GET", url_suffix="/computergroups").get("computerGroups")
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 5]
```diff
@@ -0,0 +1,1824 @@
...
+
+    def list_firewall_rules(self) -> List[Dict[str, Any]]:
+        """
+        List all firewall rules inside Trend Micro.
+
+        Returns:
+            List[Dict[str, Any]]: All firewall rules.
+        """
+
+        return self._http_request(method="GET", url_suffix="/firewallrules")["firewallRules"]
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:293]:
```suggestion
        return self._http_request(method="GET", url_suffix="/firewallrules").get("firewallRules")
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 6]
```diff
@@ -0,0 +1,1824 @@
...
+    Returns:
+        CommandResults: Command results with raw response, outputs and readable outputs.
+    """
+
+    response = client.create_computer(expand=expand, overrides=overrides, host_name=host_name,
+                                      display_name=display_name, description=description, group_id=group_id,
+                                      policy_id=policy_id, asset_importance_id=asset_importance_id,
+                                      relay_list_id=relay_list_id)
+
+    markdown = tableToMarkdown(f"Details for the new computer {response['hostName']}", response, removeNull=True,
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:709]:
```suggestion
    markdown = tableToMarkdown(f"Details for the new computer {response.get('hostName')}", response, removeNull=True,
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 7]
```diff
@@ -0,0 +1,1824 @@
...
+        expand (List[str]): The desired information about the computers.
+        overrides (bool): Whether to get the overridden properties or not.
+
+    Returns:
+        CommandResults: Command results with raw response, outputs and readable outputs.
+    """
+
+    response = client.get_computer(computer_id=computer_id, expand=expand, overrides=overrides)
+
+    markdown = tableToMarkdown(f"Details for the computer {response['hostName']}", response, removeNull=True,
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:732]:
```suggestion
    markdown = tableToMarkdown(f"Details for the computer {response.get('hostName')}", response, removeNull=True,
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 8]
```diff
@@ -0,0 +1,1824 @@
...
+    Returns:
+        CommandResults: Command results with raw response, outputs and readable outputs.
+    """
+
+    response = client.modify_computer(computer_id=computer_id, expand=expand, overrides=overrides, host_name=host_name,
+                                      display_name=display_name, description=description, group_id=group_id,
+                                      policy_id=policy_id, asset_importance_id=asset_importance_id,
+                                      relay_list_id=relay_list_id)
+
+    markdown = tableToMarkdown(f"Details for the computer {response['hostName']}", response, removeNull=True,
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:768]:
```suggestion
    markdown = tableToMarkdown(f"Details for the computer {response.get('hostName')}", response, removeNull=True,
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 9]
```diff
@@ -0,0 +1,1824 @@
...
+    Reset a certain default policy setting.
+
+    Args:
+        client (Client): The Trend Micro API client.
+
+    Returns:
+        CommandResults: Command results with raw response, outputs and readable outputs.
+    """
+
+    response = {k: v["value"] for k, v in client.list_default_policy_settings().items()}
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:1601]:
```suggestion
    response = {k: v.get("value") for k, v in client.list_default_policy_settings().items()}
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 10]
```diff
@@ -0,0 +1,1824 @@
...
+        return str(e)
+
+
+def main():
+    params = demisto.params()
+
+    use_ssl = not params.get("insecure", False)
+    use_proxy = params.get("proxy", False)
+
+    client = Client(params["server_url"], params["api_secret"], use_ssl, use_proxy)
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:1759]:
```suggestion
    client = Client(params.get("server_url"), params("api_secret"), use_ssl, use_proxy)
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 11]
```diff
@@ -0,0 +1,1824 @@
...
+                                     "trendmicro-list-policies": list_policies_command,
+                                     "trendmicro-create-policy": create_policy_command, "test-module": test_module}
+
+    error_message = ""
+
+    try:
+        command = demisto.command()
+
+        if command in commands:
+            command_function = commands[command]
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:1806]:
```suggestion
            command_function = commands.get(command)
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 12]
```diff
@@ -0,0 +1,1824 @@
...
+            return_results(command_function(client, **convert_args(command_function, demisto.args())))
+        else:
+            raise NotImplementedError(f"The command {command} does not exist on TrendMicro!")
+    except (ConnectionError, InvalidURL, InvalidSchema) as e:
+        demisto.error(traceback.format_exc())
+        error_message = f"{INVALID_URL_ERROR}\nError:\n{e}"
+    except HTTPError as e:
+        demisto.error(traceback.format_exc())
+        error_message = f"Error in API call [{e.response.status_code}]\n{e.response.json()['message']}"
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:1815]:
```suggestion
        error_message = f"Error in API call [{e.response.status_code}]\n{e.response.json().get('message')}"
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.yml [hunk 13]
```diff
@@ -0,0 +1,4051 @@
...
+    outputs:
+    - contextPath: TrendMicro.DefaultPolicySettings.name
+      description: The name of the default policy setting
+    - contextPath: TrendMicro.DefaultPolicySettings.value
+      description: The value of the default policy setting
+  dockerimage: demisto/python3:3.9.5.20070
+  runonce: false
+  subtype: python3
+  type: python
+fromversion: 5.0.0
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.yml:4051]:
```suggestion
fromversion: 6.0.0
```

File: Packs/TrendMicroDeepSecurity/TestPlaybooks/playbook-TrendMicroDeepSecurity-Test.yml [hunk 14]
```diff
@@ -0,0 +1,3003 @@
+id: Trend Micro Deep Security - Test
+version: 19
+fromversion: 5.0.0
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/TestPlaybooks/playbook-TrendMicroDeepSecurity-Test.yml:3]:
```suggestion
fromversion: 6.0.0
```

**Review by Noy-Maimon — CHANGES_REQUESTED** [review]
Thank you for your contribution! 
I requested a few minor changes to match our conventions, please have a look.

**Comment by Noy-Maimon** [comment]
@kirbles19 @ShirleyDenkberg Hi hi, can you please go over the .yml, description and readme files? Thanks

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.yml [hunk 15]
```diff
@@ -0,0 +1,4051 @@
...
+script:
+  script: "-"
+  commands:
+  - arguments:
+    - auto: PREDEFINED
+      default: false
+      defaultValue: none
+      description: Determines the computer information to include in the response.
+        Notice that the information is appended to the context data only
+      isArray: true
```
[the remark this answers is not in the document]
Comment (RazOmry) [reply Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.yml:38]: Hey! 😃 
I call to `argToList` inside the `convert_arg` function. The `convert_arg` function converts the arguments from XSOAR (strings) to their real types (one of them is list). 😄

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 16]
```diff
@@ -0,0 +1,1829 @@
...
+        demisto.error(traceback.format_exc())
+        error_message = f"{INVALID_URL_ERROR}\nError:\n{e}"
+    except HTTPError as e:
+        demisto.error(traceback.format_exc())
+        error_message = f"Error in API call [{e.response.status_code}]\n{e.response.json().get('message')}"
+    except Exception as e:
+        demisto.error(traceback.format_exc())
+        error_message = f"Failed to execute {demisto.command()} command.\nError:\n{e}"
+
+    return_error(error_message)
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:1825]:
```suggestion
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 17]
```diff
@@ -0,0 +1,1829 @@
...
+        if not command_function:
+            raise NotImplementedError(f"The command {command} does not exist on TrendMicro!")
+        else:
+            return_results(command_function(client, **convert_args(command_function, demisto.args())))
+    except (ConnectionError, InvalidURL, InvalidSchema) as e:
+        demisto.error(traceback.format_exc())
+        error_message = f"{INVALID_URL_ERROR}\nError:\n{e}"
+    except HTTPError as e:
+        demisto.error(traceback.format_exc())
+        error_message = f"Error in API call [{e.response.status_code}]\n{e.response.json().get('message')}"
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:1820]:
```suggestion
        error_message = f"Error in API call [{e.response.status_code}]\n{e.response.json().get('message')}"
        return_error(error_message)
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 18]
```diff
@@ -0,0 +1,1829 @@
...
+    except Exception as e:
+        demisto.error(traceback.format_exc())
+        error_message = f"Failed to execute {demisto.command()} command.\nError:\n{e}"
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:1823]:
```suggestion
        error_message = f"Failed to execute {demisto.command()} command.\nError:\n{e}"
        return_error(error_message)
```

File: Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py [hunk 19]
```diff
@@ -0,0 +1,1829 @@
...
+        command = demisto.command()
+
+        command_function = commands.get(command)
+        if not command_function:
+            raise NotImplementedError(f"The command {command} does not exist on TrendMicro!")
+        else:
+            return_results(command_function(client, **convert_args(command_function, demisto.args())))
+    except (ConnectionError, InvalidURL, InvalidSchema) as e:
+        demisto.error(traceback.format_exc())
+        error_message = f"{INVALID_URL_ERROR}\nError:\n{e}"
```
Comment (Noy-Maimon) [inline Packs/TrendMicroDeepSecurity/Integrations/TrendMicroDeepSecurity/TrendMicroDeepSecurity.py:1817]:
```suggestion
        error_message = f"{INVALID_URL_ERROR}\nError:\n{e}"
        return_error(error_message)
```

**Review by Noy-Maimon — CHANGES_REQUESTED** [review]

**Comment by Noy-Maimon** [comment]
@RazOmry Please see the suggested changes as discussed via slack.
Once they are committed, let's schedule a demo. Please let me know when are you available.
Thanks!

**Comment by ShirleyDenkberg** [comment]
@Noy-Maimon Doc review completed.  Since I did not have permission to edit the files directly in this PR, I sent the edited files to you by Slack.  Please replace the files in this PR.  The Integration README needs to be regenerated.

**Pull request merged**